### PR TITLE
TCP support

### PR DESCRIPTION
**Pull request by mbiegert**
Implements a rudimentary tcp implementation of the protocol.
Also adds a little testserver file to test the implementation.
Also adds a quick and dirty tup file to build the testserver.
Also I'm tired...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,3 +83,11 @@ Note that this is not yet representative for the existing code.
 - Add compatibility with Windows, macOS, iOS, Android
```
Comment (samuelsadok) [inline README.md]: you can add TCP to Compatibility list

File: cpp/posix_tcp.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#include <unistd.h>
+#include <thread>
+#include <future>
+
+#include <fibre/protocol.hpp>
+
+
+#define TCP_RX_BUF_LEN	512
+
+class TCPBytesSender : public StreamSink {
```
Comment (samuelsadok) [inline cpp/posix_tcp.cpp]: `TCPStreamSink` would be more consistent

File: cpp/posix_tcp.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    int process_bytes(const uint8_t* buffer, size_t length, size_t& processed_bytes) {
+        int bytes_sent = send(_socket_fd, buffer, length, 0);
+        processed_bytes = (bytes_sent == -1) ? 0 : bytes_sent;
+        return (bytes_sent == -1) ? -1 : 0;
+    }
+
+    size_t get_free_space() { return SIZE_MAX; }
+
+private:
+    int _socket_fd;
```
Comment (samuelsadok) [inline cpp/posix_tcp.cpp]: the [Google Naming Convention](https://google.github.io/styleguide/cppguide.html#Variable_Names) would be `socket_fd_`

File: cpp/posix_tcp.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    BidirectionalPacketBasedChannel channel(endpoints, n_endpoints, packet2stream);
+
+    StreamToPacketSegmenter stream2packet(channel);
+
+    // now listen for it
+    for (;;) {
+        memset(buf, 0, sizeof(buf));
+        // returns as soon as there is some data
+        ssize_t n_received = recv(sock_fd, buf, sizeof(buf), 0);
+        printf("recvd something: %d\n", n_received);
```
Comment (samuelsadok) [inline cpp/posix_tcp.cpp]: this is probably not intended (or should be wrapped in `#ifdef DEBUG_PROTOCOL`)

File: cpp/posix_tcp.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+bool future_is_ready(std::future<T>& t){
+    return t.wait_for(std::chrono::seconds(0)) == std::future_status::ready;
+}
+
+int serve_on_tcp(const Endpoint endpoints[], size_t n_endpoints, unsigned int port) {
+    struct sockaddr_in si_me, si_other;
+    int s;
+
+
+    if ((s=socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) == -1)
```
Comment (samuelsadok) [inline cpp/posix_tcp.cpp]: It seems if you change this to `AF_INET6` we automatically magically support both IPv4 and IPv6

File: cpp/posix_tcp.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        return -1;
+
+    listen(s, 128); // make this socket a passive socket
+    std::vector<std::future<int>> serv_pool;
+    for (;;) {
+        memset(&si_other, 0, sizeof(si_other));
+
+        socklen_t silen = sizeof(si_other);
+        printf("enter blocking accept\n");
+        int client_portal_fd = accept(s, reinterpret_cast<sockaddr *>(&si_other), &silen); // blocking call
```
Comment (samuelsadok) [inline cpp/posix_tcp.cpp]: maybe add a TODO to note that we should add a compile-time or runtime limit for the number of concurrent connections

File: python/fibre/tcp_transport.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    # convert deadline to seconds (floating point)
+    deadline = None if deadline is None else max(deadline - time.monotonic(), 0)
+    self.sock.settimeout(deadline)
+    try:
+      data = self.sock.recv(n_bytes, socket.MSG_WAITALL) # receive n_bytes
+      return data
+    except TimeoutError:
+      # if we got a timeout data will still be none, so we call recv again
+      # this time in non blocking state and see if we can get some data
+      return self.get_bytes(n_bytes, time.monotonic()-1)
```
Comment (samuelsadok) [inline python/fibre/tcp_transport.py]: `return self.sock.recv(n_bytes, socket.MSG_DONTWAIT)`

File: cpp/__test_server.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
+
```
Comment (samuelsadok) [inline cpp/__test_server.cpp]: please move this to `tests/test_server.cpp` (along with the Tupfile)

Comment (mbiegert) [reply cpp/__test_server.cpp]: no, I argue for the tupfile to be able to build the project to your needs. Let me explain my *vision* using lightd: lightd has a tupfile in the top level directory. There you include fibre as a subdirectory and tup will automatically build the fibre and use it in the ligthd project without the need to specify every fibre file necessary for this. I'm currently refactoring the tupfile and will remove it from the PR for the time being.

File: cpp/posix_tcp.cpp (hunk 3 above)
Comment (mbiegert) [reply cpp/posix_tcp.cpp]: maybe add a link saying this to the developer part of the readme.

File: cpp/posix_tcp.cpp (hunk 5 above)
Comment (mbiegert) [reply cpp/posix_tcp.cpp]: it appears you're right, will check that

File: cpp/posix_tcp.cpp (hunk 3 above)
Comment (samuelsadok) [reply cpp/posix_tcp.cpp]: yes, will do after merging this to avoid merge conflicts

**Pull request merged**